File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/ArchiveMetadataWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.libxmtp.ArchiveElement
import org.xmtp.android.library.libxmtp.ArchiveMetadata

class ArchiveMetadataWrapper {
    companion object {
        val gson: Gson = GsonBuilder().create()
        private fun encodeToObj(metadata: ArchiveMetadata): Map<String, Any?> {
            return mapOf(
                "archiveVersion" to metadata.archiveVersion,
                "elements" to try {
                    metadata.elements.map { getArchiveElementString(it) }
                } catch (e: Exception) {
                    // By default we archive everything
                    listOf("message", "consent")
                },
                "exportedAtNs" to metadata.exportedAtNs,
                "startNs" to metadata.startNs,
                "endNs" to metadata.endNs
            )
        }

        private fun getArchiveElementString(element: ArchiveElement): String {
            return when (element) {
                ArchiveElement.CONSENT -> "consent"
                ArchiveElement.MESSAGES -> "message"
                else -> throw XMTPException("Invalid archive element: $element")
            }
        }

        /** Returns metadata as a map for embedding in another JSON object (avoids double JSON encoding). */
        fun encodeToMap(metadata: ArchiveMetadata?): Map<String, Any?> {
            return if (metadata != null) {
                encodeToObj(metadata)
            } else {
                mapOf(
                    "archiveVersion" to 0u,
                    "elements" to listOf("messages", "consent"),
                    "exportedAtNs" to 0L,
                    "startNs" to null,
                    "endNs" to null
                )
            }
        }

        fun encode(metadata: ArchiveMetadata?): String {
            return gson.toJson(encodeToMap(metadata))
        }
    }
}

File: ios/Wrappers/ArchiveMetadataWrapper.swift
import Foundation
import XMTP

// Wrapper around XMTP.ArchiveMetadata to allow passing these objects back into react native.
struct ArchiveMetadataWrapper {
    static func encodeToObj(_ metadata: XMTP.ArchiveMetadata) throws -> [String: Any] {
        return [
            "archiveVersion": metadata.archiveVersion,
            "elements": getArchiveElementStrings(metadata.elements),
            "exportedAtNs": metadata.exportedAtNs,
            "startNs": metadata.startNs,
            "endNs": metadata.endNs
        ]
    }
    
    private static func getArchiveElementStrings(_ elements: [XMTP.ArchiveElement]) -> [String] {
        do {
            return try elements.map { try getArchiveElementString($0) }
        } catch {
            // By default we archive everything
            return ["messages", "consent"]
        }
    }
    
    private static func getArchiveElementString(_ element: XMTP.ArchiveElement) throws -> String {
        switch element {
        case .messages:
            return "messages"
        case .consent:
            return "consent"
        default:
            throw WrapperError.encodeError("Invalid archive element: \(element)")
        }
    }
    
    static func encode(_ metadata: XMTP.ArchiveMetadata?) throws -> String {
        let obj: [String: Any]
        
        if let metadata = metadata {
            obj = try encodeToObj(metadata)
        } else {
            // Create default metadata object when null
            obj = [
                "archiveVersion": 0,
                "elements": ["messages", "consent"],
                "exportedAtNs": 0,
                "startNs": NSNull(),
                "endNs": NSNull()
            ]
        }
        
        let data = try JSONSerialization.data(withJSONObject: obj)
        guard let result = String(data: data, encoding: .utf8) else {
            throw WrapperError.encodeError("could not encode archive metadata")
        }
        return result
    }
}